c/tree.c

```diff
@@ -589,36 +592,113 @@ SV *merge_hashes_for_keys(MMDBW_tree_s *tree, const char *const key_from,
+{
+    HV *hash_from = (HV *)SvRV(from);
+    HV *hash_into = (HV *)SvRV(into);
     HV *hash_new = newHV();
 
-    merge_hash(hash_into, hash_new);
-    merge_hash(hash_from, hash_new);
+    merge_new_from_hash_into_hash(tree, hash_from, hash_new);
+    merge_new_from_hash_into_hash(tree, hash_into, hash_new);
 
     return newRV_noinc((SV *)hash_new);
 }
 
-LOCAL void merge_hash(HV *from, HV *to)
+// Note: unlike the other merge functions, this does _not_ replace existing
+// values.
+LOCAL void merge_new_from_hash_into_hash(MMDBW_tree_s *tree, HV *from, HV *to)
 {
     (void)hv_iterinit(from);
     HE *he;
     while (NULL != (he = hv_iternext(from))) {
         STRLEN key_length;
         const char *const key = HePV(he, key_length);
-        U32 hash = HeHASH(he);
+        U32 hash = 0;
+        SV *value = HeVAL(he);
         if (hv_exists(to, key, key_length)) {
-            continue;
+            if (tree->merge_strategy == MMDBW_MERGE_STRATEGY_RECURSE) {
+                SV **existing_value = hv_fetch(to, key, key_length, 0);
+                if (existing_value == NULL) {
+                    // This should never happen as we just did an hv_exists
+                    croak("Received an unexpected NULL from hv_fetch");
+                }
+                value = merge_values(tree, value, *existing_value);
+            } else {
+                continue;
+            }
+        } else {
+            hash = HeHASH(he);
+            SvREFCNT_inc_simple_void_NN(value);
         }
 
-        SV *value = HeVAL(he);
-        SvREFCNT_inc_simple_void_NN(value);
         (void)hv_store(to, key, key_length, value, hash);
     }
 
     return;
 }
 
+LOCAL SV * merge_values(MMDBW_tree_s *tree, SV *from, SV *into)
+{
+    if (SvROK(from) != SvROK(into)) {
+        croak("Attempt to merge a reference value and non-refrence value");
+    }
+
+    if (!SvROK(from)) {
+        // If the two values are scalars, we prefer the one in the hash being
+        // inserted.
+        SvREFCNT_inc_simple_void_NN(from);
+        return from;
+    }
+
+    if (SvTYPE(SvRV(from)) == SVt_PVHV && SvTYPE(SvRV(into)) == SVt_PVHV) {
+        return merge_hashes(tree, from, into);
+    }
+
+    if (SvTYPE(SvRV(from)) == SVt_PVAV && SvTYPE(SvRV(into)) == SVt_PVAV) {
+        return merge_arrays(tree, from, into);
+    }
+
+    croak("Only arrayrefs, hashrefs, and scalars can be merged.");
+}
+
+LOCAL SV * merge_arrays(MMDBW_tree_s *tree, SV *from, SV *into)
+{
+    AV *from_array = (AV *)SvRV(from);
+    AV *into_array = (AV *)SvRV(into);
+
+    // Note that av_len() is really the index of the last element. In newer
+    // Perl versions, it is also called av_top_index() or av_tindex()
+    SSize_t from_top_index = av_len(from_array);
+    SSize_t into_top_index = av_len(into_array);
+
+    SSize_t new_top_index = from_top_index >
+                            into_top_index ? from_top_index : into_top_index;
+
+    AV *new_array = newAV();
```

Comment:
Maybe we should be pre-allocating the size of the AV we're creating here with av_extend and then use av_store later rather than av_push?  This would avoid re-allocating arrays as stuff is pushed into them / wasting space with bigger arrays than needed